### DISCUSSION TITLE
Python API exit handling

### DISCUSSION THREAD

**Discussion by marioaag** [avocado-framework/avocado#4514] [Q&A] [answered]
Hi All,

Using the Python API noticed that it is ending in un expected way, it is not thronging an exception and neither continue with the execution of the rest of the code that is after the call of the run() method, it is jumping directly to the finally statement.

Example:
```
    with Job(config=JOB_CONFIG, test_suites=test_suites) as job:
        job.run()

    logging.info("****To Be Implemented****")
```

In this simple example the expectation is that the logging get executed but it is not.

Is there a way to continue with the normal flow of the script where the avocado API is being used?

**Comment by beraldoleal** [comment]
Hi @marioaag,  what are your tests? maybe your test is tucked by some reason? Could you please paste a complete example with a reproducer?

**Reply by marioaag** [reply]
Here is an example

Script calling the API example:
```
import logging
from avocado.core.job import Job
from avocado.core.suite import TestSuite
from avocado.core.nrunner import Runnable

# Generating test paths
JOB_CONFIG = {'run.test_runner': 'nrunner', 'core.show': ['app']}
test_suites = []
tests = []
tests_name = ['/Users/marioalvarado/git/iLabConfig/testcases/check_host_status.py:CheckHostStatus.test_os_info', '/Users/marioalvarado/git/iLabConfig/testcases/check_host_status.py:CheckHostStatus.test_file_system_healt']

for test_name in tests_name:
    tests.append(Runnable('avocado-instrumented', test_name))
    logging.info(f'Test references to be used: {test_name}')

# Avocado Framework Test Suites Generation
test_suite = TestSuite(name=test_name,
                                      config={'run.test_parameters': [['host_ip', '300']]},
                                      tests=tests,
                                      job_config=JOB_CONFIG)
test_suites.append(test_suite)

# Executing tests with Avocado Framework New Runner
with Job(config=JOB_CONFIG, test_suites=test_suites) as job:
    job.run()

logging.info("****To Be Implemented****")
```

Tests implemented example:
```
from time import sleep
from avocado import Test

class CheckHostStatus(Test):
    """ """

    def test_os_info(self):
        """Retrieve OS information using ssh connection"""
        # Retrieving required parameters
        host_ip = self.params.get('host_ip', default=111)
        sleep(10)
        self.log.info(f'Host IP: {host_ip}')
        self.log.debug("Debug Test")
        self.log.info("Info Test")

    def test_file_system_healt(self):
        """Retrieve OS information using ssh connection"""
        # Retrieving required parameters
        host_ip = self.params.get('host_ip', default=112)
        sleep(10)
        self.log.info(f'Host IP: {host_ip}')
```

Here the behavior to check is that the logging with the message "****To Be Implemented****" is not getting executed

**Reply by clebergnu** [reply, marked as answer]
> Here is an example
> 
> Script calling the API example:
> 
> ```
> import logging
> from avocado.core.job import Job
> from avocado.core.suite import TestSuite
> from avocado.core.nrunner import Runnable
> 
> # Generating test paths
> JOB_CONFIG = {'run.test_runner': 'nrunner', 'core.show': ['app']}
> test_suites = []
> tests = []
> tests_name = ['/Users/marioalvarado/git/iLabConfig/testcases/check_host_status.py:CheckHostStatus.test_os_info', '/Users/marioalvarado/git/iLabConfig/testcases/check_host_status.py:CheckHostStatus.test_file_system_healt']
> 
> for test_name in tests_name:
>     tests.append(Runnable('avocado-instrumented', test_name))
>     logging.info(f'Test references to be used: {test_name}')
> 
> # Avocado Framework Test Suites Generation
> test_suite = TestSuite(name=test_name,
>                                       config={'run.test_parameters': [['host_ip', '300']]},
>                                       tests=tests,
>                                       job_config=JOB_CONFIG)
> test_suites.append(test_suite)
> 
> # Executing tests with Avocado Framework New Runner
> with Job(config=JOB_CONFIG, test_suites=test_suites) as job:
>     job.run()
> 
> logging.info("****To Be Implemented****")
> ```
> 
> Tests implemented example:
> 
> ```
> from time import sleep
> from avocado import Test
> 
> class CheckHostStatus(Test):
>     """ """
> 
>     def test_os_info(self):
>         """Retrieve OS information using ssh connection"""
>         # Retrieving required parameters
>         host_ip = self.params.get('host_ip', default=111)
>         sleep(10)
>         self.log.info(f'Host IP: {host_ip}')
>         self.log.debug("Debug Test")
>         self.log.info("Info Test")
> 
>     def test_file_system_healt(self):
>         """Retrieve OS information using ssh connection"""
>         # Retrieving required parameters
>         host_ip = self.params.get('host_ip', default=112)
>         sleep(10)
>         self.log.info(f'Host IP: {host_ip}')
> ```
> 
> Here the behavior to check is that the logging with the message "****To Be Implemented****" is not getting executed

@marioaag it probably *is* getting executed, but not being shown.  This is happening because the job will configure the `logging` module, and may be failing to reset it completely.  For instance, the following works:

```python
# Executing tests with Avocado Framework New Runner
with Job(config=JOB_CONFIG, test_suites=test_suites) as job:
    job.run()
    job.log.info("****To Be Implemented****")
```

I'll double check where the issue is, and let you know.

**Reply by marioaag** [reply]
you are right it is related with the logging module configuration! I did a check replacing with print() and it works in all the different places.

**Reply by marioaag** [reply]
@clebergnu, is there any workaround to avoid getting the log handlers configured previously to the avocado execution unconfigured?
